network-config-analyzer/IstioNetworkPolicy.py

```diff
@@ -0,0 +1,188 @@
+
+    def allowed_connections(self, from_peer, to_peer, is_ingress):
+        """
+        Evaluate the set of connections this policy allows/denies/passes between two peers
+        :param Peer.Peer from_peer: The source peer
+        :param Peer.Peer to_peer:  The target peer
+        :param bool is_ingress: whether we evaluate ingress rules only or egress rules only
+        :return: A PolicyConnections object containing sets of allowed/denied/pass connections
+        :rtype: PolicyConnections
+        """
+
+        # TODO: currently not handling egress, istio authorization policies have no egress rules
+        if not is_ingress:
+            return PolicyConnections(False, ConnectionSet(True))
+
+        captured = to_peer in self.selected_peers
+        if not captured:
+            return PolicyConnections(False)
+
+        allowed_conns = ConnectionSet()
+        denied_conns = ConnectionSet()
+
+        rules = self.ingress_rules
+        collected_conns = allowed_conns if self.action == IstioNetworkPolicy.ActionType.Allow else denied_conns
+        for rule in rules:
+            if from_peer in rule.peer_set:
+                rule_conns = rule.connections
+                collected_conns |= rule_conns
+
+        return PolicyConnections(True, allowed_conns, denied_conns)
+
+    def referenced_ip_blocks(self):
+        """
+        :return: A set of all ipblocks referenced in one of the policy rules (one Peer object per one ip range)
+        :rtype: Peer.PeerSet
+        """
+        res = PeerSet()
+        for rule in self.ingress_rules:
+            for peer in rule.peer_set:
+                if isinstance(peer, IpBlock):
+                    res |= peer.split()
+        return res
+
+    def has_empty_rules(self, config_name=''):
+        """
+        Checks whether the policy contains empty rules (rules that do not select any peers)
+        :param str config_name: (optional) the name of the NetworkConfig object
+        :return: A list of strings describing the emptiness + two sets of indexes of empty ingress/egress rules
+        :rtype: list[str], set, set
+        """
+        emptiness_explanation = []
+        empty_ingress_rules = set()
+        full_name = self.full_name(config_name)
+        for rule_index, ingress_rule in enumerate(self.ingress_rules, start=1):
+            if not ingress_rule.peer_set:
+                emptiness = f'Ingress rule no. {rule_index} in AuthorizationPolicy {full_name} does not select any pods'
+                emptiness_explanation.append(emptiness)
+                empty_ingress_rules.add(rule_index)
+
+        return emptiness_explanation, empty_ingress_rules, set()
+
+    def clone_without_rule(self, rule_to_exclude, ingress_rule):
+        """
+        Makes a copy of 'self' without a given policy rule
+        :param IstioPolicyRule rule_to_exclude: The one rule not to include in the copy
+        :param bool ingress_rule: Whether the rule is an ingress or egress rule
```

Comment:
```suggestion
    def clone_without_rule(self, rule_to_exclude, _ingress_rule):
        """
        Makes a copy of 'self' without a given policy rule
        :param IstioPolicyRule rule_to_exclude: The one rule not to include in the copy
        :param bool _ingress_rule: Whether the rule is an ingress or egress rule
```
The underscore marks the variable is not used

Reply:
done